### PR TITLE
Deprecate Data.Array.Foreign module

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

File: core/src/Streamly/Data/Array/Unboxed.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+{-# OPTIONS_GHC -Wno-deprecations #-}
+#include "inline.hs"
+
+-- |
+-- Module      : Streamly.Data.Array.Foreign
```
Comment (adithyaov) [inline core/src/Streamly/Data/Array/Unboxed.hs:5]: `Streamly.Data.Array.Unboxed`

File: core/src/Streamly/Data/Array/Unboxed.hs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+-- Copyright   : (c) 2019 Composewell Technologies
```
Comment (adithyaov) [inline core/src/Streamly/Data/Array/Unboxed.hs:6]: `2022`

File: core/src/Streamly/Data/Array/Unboxed.hs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+--
+-- License     : BSD3
```
Comment (adithyaov) [inline core/src/Streamly/Data/Array/Unboxed.hs:8]: `BSD-3-Clause`

File: core/src/Streamly/Data/Array/Foreign.hs [hunk 4]
```diff
@@ -99,6 +100,6 @@ import Streamly.Internal.Data.Array.Foreign as A
 -- >>> :set -XFlexibleContexts
 -- >>> :set -package streamly
 -- >>> import Streamly.Prelude (SerialT)
--- >>> import Streamly.Data.Array.Foreign (Array)
+-- >>> import Streamly.Data.Array.Unboxed (Array)
```
Comment (adithyaov) [inline core/src/Streamly/Data/Array/Foreign.hs:103]: Remove the module documentation. Point it to `Data.Array.Unboxed`

File: core/src/Streamly/Data/Array/Unboxed.hs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+-- the operating system.
+--
+-- Arrays in this module are chunks of pinned memory that hold a sequence of
+-- 'Storable' values of a given type, they cannot store non-serializable data
+-- like functions.  Once created an array cannot be modified.  Pinned memory
+-- allows efficient buffering of long lived data without adding any impact to
+-- GC. One array is just one pointer visible to GC and it does not have to be
+-- copied across generations.  Moreover, pinned memory allows communication
+-- with foreign consumers and producers (e.g. file or network IO) without
+-- copying the data.
```
Comment (adithyaov) [inline core/src/Streamly/Data/Array/Unboxed.hs:24]: For the time being, add an XXX comment stating to change the documentation and open an issue about it.

File: core/streamly-core.cabal [hunk 6]
```diff
@@ -333,6 +333,7 @@ library
                      , Streamly.Data.Fold
                      , Streamly.Data.Fold.Tee
                      , Streamly.Data.Array.Foreign
+                     , Streamly.Data.Array.Unboxed
```
Comment (adithyaov) [inline core/streamly-core.cabal:336]: Move `Streamly.Data.Array.Foreign` under a `-- Deprecated` comment. See `streamly.cabal`

File: streamly.cabal [hunk 7]
```diff
@@ -650,6 +651,7 @@ library
                      , Streamly.Data.Fold
                      , Streamly.Data.Fold.Tee
                      , Streamly.Data.Array.Foreign
+                     , Streamly.Data.Array.Unboxed
```
Comment (adithyaov) [inline streamly.cabal:654]: Move `Streamly.Data.Array.Foreign` under a `-- Deprecated` comment.

**Comment by harendra-kumar** [comment]
We need to do the following:

* git move `Streamly.Internal.Data.Array.Foreign.*` to `Streamly.Internal.Data.Array.Unboxed.*`
* Create `Streamly.Data.Array.Unboxed` as an alternative to `Streamly.Data.Array.Foreign`
* Deprecate `Streamly.Data.Array.Foreign`

Note that streamly-core package must not contain the deprecated `Streamly.Data.Array.Foreign` module, it should be only in streamly package.

**Pull request closed without merging**